src/rgw/rgw_qos.h

```diff
@@ -0,0 +1,246 @@
+      int64_t read_bw = tenant.max_read_bytes * time_in_ms;
+      it->second.read[RGW_QOS_OPS] = std::min(tenant.max_read_ops * fixed_float, read_ops + it->second.read[RGW_QOS_OPS]);
+      it->second.read[RGW_QOS_BW] = std::min(tenant.max_read_bytes * fixed_float, read_bw + it->second.read[RGW_QOS_BW]);
+
+      it->second.write[RGW_QOS_OPS] = std::min(tenant.max_write_ops * fixed_float, write_ops + it->second.write[RGW_QOS_OPS]);
+      it->second.write[RGW_QOS_BW] = std::min(tenant.max_write_bytes * fixed_float, write_bw + it->second.write[RGW_QOS_BW]);
+
+    } else if(it->second.first_run)
+      {
+        it->second.write[RGW_QOS_OPS] = tenant.max_write_ops * fixed_float;
+        it->second.write[RGW_QOS_BW] = tenant.max_write_bytes * fixed_float;
+        it->second.read[RGW_QOS_OPS] = tenant.max_read_ops * fixed_float;
+        it->second.read[RGW_QOS_BW] = tenant.max_read_bytes * fixed_float;
+        it->second.ts = curr_timestamp;
+        it->second.first_run = false;
+      }
+      return true;
+    }
+
+  bool increase_entry_write(hash_map::iterator& it, int64_t ops_limit, int64_t bw_limit) {
+    //check if tenants did not reach their bw or ops limits and that the limits are not 0 (which is unlimited)
+    if(((it->second.write_ops() - 1 < 0) && (ops_limit > 0)) ||
+      (it->second.write_bytes() < 0 && bw_limit > 0))
+    {
+      return true;
+    }
+
+    // we don't want to reduce ops' tokens if we've rejected it.
+    it->second.write[RGW_QOS_OPS] -= fixed_float_rgw_ratelimit;
+    return false;
+  }
+
+  public:
+    QosDatastruct(const QosDatastruct&) = delete;
+    QosDatastruct& operator =(const QosDatastruct&) = delete;
+    QosDatastruct(QosDatastruct&&) = delete;
+    QosDatastruct& operator =(QosDatastruct&&) = delete;
+    QosDatastruct() {};
+    QosDatastruct(std::atomic_bool *replacing, std::condition_variable *cv) {
+      // prevent rehash, so no iterators invalidation
+      qos_entries.max_load_factor(1000);
+      this->replacing = replacing;
+      this->cv = cv;
+    };
+    bool increase_entry(const char *method, std::string tenant, ceph::coarse_real_time curr_timestamp, const RGWQoSInfo& qos_info) {
+      if(tenant.empty() || tenant.length() == 1 || !qos_info.enabled)
+        return false;
+      bool is_read = is_read_op(method);
+      // it is possible that some requests will have the same ts, because we are using coarse real time
+      auto it = find_or_create(tenant);
+      auto curr_ts = curr_timestamp.time_since_epoch();
+      std::unique_lock lock(it->second.ts_lock);
+      increase_tokens(it, curr_ts, qos_info);
+      if(is_read)
+        return increase_entry_read(it, qos_info.max_read_ops, qos_info.max_read_bytes);
+      return increase_entry_write(it, qos_info.max_write_ops, qos_info.max_write_bytes);
+    }
+    bool increase_bw(const char *method, std::string tenant, int64_t amount, const RGWQoSInfo& info) {
+      if(tenant.empty() || tenant.length() == 1 || !info.enabled)
```

Comment:
rgw already has a very specific meaning for "tenant", so we should find a different term in the context of rate limiting. here, i would probably just call this the "key" since it's what we use to access the map